README.md
# pyd2v

[![License](https://img.shields.io/github/license/rlaphoenix/pyd2v)](https://github.com/rlaphoenix/pyd2v/blob/master/LICENSE)
[![Build status](https://github.com/rlaphoenix/pyd2v/actions/workflows/ci.yml/badge.svg)](https://github.com/rlaphoenix/pyd2v/actions/workflows/ci.yml)
[![Python version](https://img.shields.io/pypi/pyversions/pyd2v)](https://pypi.python.org/pypi/pyd2v)
[![PyPI version](https://img.shields.io/pypi/v/pyd2v)](https://pypi.python.org/pypi/pyd2v)
[![Contributors](https://img.shields.io/github/contributors/rlaphoenix/pyd2v)](https://github.com/rlaphoenix/pyd2v/graphs/contributors)
[![DeepSource issues](https://deepsource.io/gh/rlaphoenix/pyd2v.svg/?label=active+issues)](https://deepsource.io/gh/rlaphoenix/pyd2v)

A Python Parser for DGMPGDec's D2V Project Files.

## Installation

    pip install --user pyd2v

## Building

### Dependencies

- [Python](https://python.org/downloads) (v3.7 or newer)
- [Poetry](https://python-poetry.org/docs) (latest recommended)

### Steps

1. `git clone https://github.com/rlaphoenix/pyd2v`
2. `cd pyd2v`
3. `poetry config virtualenvs.in-project true` (optional, but recommended)
4. `poetry install`
5. `d2v --help`

For further information on using the Poetry virtual environment, see
[Using your virtual environment](https://python-poetry.org/docs/basic-usage/#using-your-virtual-environment).

## Usage

This project can be used in Python Scripts by importing `pyd2v`, or in your terminal by calling `d2v`.

### Scripting

```py
from pyd2v import D2V

# ...

d2v = D2V(filename="C:/Users/phoenix/Videos/1998_home_video.d2v")
print(d2v)  # d2v object basic information, e.g. version and settings
print(d2v.videos)  # input video files
print(d2v.settings["Frame_Rate"])  # frame rate
print(d2v.data[0])  # print first frame data
```

### Executable

For help information run `d2v --help`. Here's some examples of using the d2v executable:

#### Getting a Setting like the Video Resolution:

```shell
$ d2v "D:\...\The.IT.Crowd.S01E01.Yesterdays.Jam.PAL.DVD.DD.2.0.MPEG-2.REMUX.d2v" settings.Picture_Size
[720, 576]
```

#### Parse the entire D2V to a JSON Pickle:

```shell
$ d2v "D:\...\The.IT.Crowd.S01E01.Yesterdays.Jam.PAL.DVD.DD.2.0.MPEG-2.REMUX.d2v" --json
{"py/object": "pyd2v.d2v.D2V", "path": {"py/reduce": [{"py/type": "pathlib.WindowsPath"}, {"py/tuple": ["D:\\", "...",
"The.IT.Crowd.S01E01.Yesterdays.Jam.PAL.DVD.DD.2.0.MPEG-2.REMUX.d2v"]}]}, "version": 16, "videos": [
"D:\\...\\The.IT.Crowd.S01E01.Yesterdays.Jam.PAL.DVD.DD.2.0.MPEG-2.REMUX.mpeg"], "settings": {"Stream_Type": 0,
"MPEG_Type": 2, "iDCT_Algorithm": 5, "YUVRGB_Scale": 1, "Luminance_Filter": {"Gamma": 0, "Offset": 0}, "Clipping":
[0, 0, 0, 0], "Aspect_Ratio": "16:9", "Picture_Size": [720, 576], "Field_Operation": 2, ... ...
```

#### Parse only the D2V Headers:

```shell
$ d2v "D:\...\The.IT.Crowd.S01E01.Yesterdays.Jam.PAL.DVD.DD.2.0.MPEG-2.REMUX.d2v" --pretty
<D2V path=WindowsPath('D:/.../The.IT.Crowd.S01E01.Yesterdays.Jam.PAL.DVD.DD.2.0.MPEG-2.REMUX.d2v') version=16,
data_type='100.00% VIDEO\n', settings={'Stream_Type': 0, 'MPEG_Type': 2, 'iDCT_Algorithm': 5, 'YUVRGB_Scale': 1,
'Luminance_Filter': {'Gamma': 0, 'Offset': 0}, 'Clipping': [0, 0, 0, 0], 'Aspect_Ratio': '16:9', 'Picture_Size':
[720, 576], 'Field_Operation': 2, 'Frame_Rate': ['25000', [25, 1]], 'Location': {'StartFile': '0', 'StartOffset': '0',
'EndFile': '0', 'EndOffset': '67d1f'}}>
```

### D2V Properties

D2V objects contain the following properties:

- version: D2V version, `16` is currently the latest for the original DGIndex which was last updated in 2010.
- videos: List of the video file paths that were indexed by DGIndex. It will be just a filename if "Use Full Paths" was
  disabled in DGIndex.
- settings: Will return various user-provided and auto-evaluated settings based on input data. More information on
  Settings below.
- data: Indexing data of the MPEG video stream, Each entry is of an I frame which will describe the following non-I
  frames up to the next I frame.
- data_type: What type of video is most previlent, e.g. `88.4% FILM`, `PAL`, `99.9% NTSC`.

### Settings

| Auto-evaluated Settings | Possible Values                        | Description                                                                        |
| ----------------------- | -------------------------------------- | ---------------------------------------------------------------------------------- |
| Stream_Type             | 0=Elementary Stream                    | Defines the type of MPEG stream.                                                   |
|                         | 1=Program Stream                       |                                                                                    |
|                         | 2=Transport Stream                     |                                                                                    |
|                         | 3=PVA Stream                           |                                                                                    |
| Transport_Packet_Size   | [188, 204]                             | Specifies the size in bytes of the transport packets. Used only for Stream_Type=2. |
| MPEG_Type               | 1=MPEG-1, 2=MPEG-2                     | Defines the type of MPEG stream.                                                   |
| Aspect_Ratio            | MPEG-2: "1:1", "4:3", "16:9", "2.21:1" | Defines the Aspect Ratio of the video specified in the MPEG stream.                |
|                         | MPEG-1: "1:1", 0.6735, ["16:9", 625],  |                                                                                    |
|                         | 0.7615, 0.8055, ["16:9", 525], 0.8935, |                                                                                    |
|                         | ["4:3", 625], 0.9815, 1.0255, 1.0695,  |                                                                                    |
|                         | ["4:3", 525], 1.575, 1.2015            |                                                                                    |
| Picture_Size            | [width, height]                        | Defines the size of the video _after_ clipping has been applied.                   |
| Frame_Rate              | rate [num, den]                        | 'rate' defines output framerate \* 1000.                                           |

| User-specified Settings | Possible Values                           | Description                                                              |
| ----------------------- | ----------------------------------------- | ------------------------------------------------------------------------ |
| MPEG2_Transport_PID     | {Video, Audio, RCR}                       | Selects the video/audio PIDs to be decoded. Used only for Stream_Type=2. |
| iDCT_Algorithm          | 1=32-bit MMX                              | Defines the iDCT DGDecode will use to decode this video                  |
|                         | 2=32-bit SSEMMX                           |                                                                          |
|                         | 3=32-bit SSE2MMX                          |                                                                          |
|                         | 4=64-bit Floating Point                   |                                                                          |
|                         | 5=64-bit IEEE-1180 Reference              |                                                                          |
|                         | 6=32-bit SSEMMX (Skal)                    |                                                                          |
|                         | 7=32-bit Simple MMX (XviD)                |                                                                          |
| YUVRGB_Scale            | 0=TV Scale                                | Defines the range DGDecode will use if RGB conversion is requested.      |
|                         | 1=PC Scale                                |                                                                          |
| Luminance_Filter        | {Gamma, Offset} (range of +/- 256)        | Defines values for DGIndex's Luminance_Filter.                           |
| Clipping                | [ClipLeft,ClipRight,ClipTop,ClipBottom]   | Defines values for Cropping lines of video.                              |
| Field_Operation         | 0=Honor Pulldown Flags                    | Defines values for Field Operation.                                      |
|                         | 1=Force Film                              |                                                                          |
|                         | 2=Ignore Pulldown Flags                   |                                                                          |
| Location                | {StartFile,StartOffset,EndFile,EndOffset} | Defines start and end points for the video selection range.              |

## License

[MIT License](LICENSE)


HISTORY.md
# Release History

## 1.4.0 (~~1.3.1~~)

Previously released as 1.3.1 when it should be 1.4.0 as it has a breaking change.

### ⚠️Breaking Changes

- Dropped support for Python 3.6.

### Improvements

- Listed the Version in the CLI help information.
- Updated the dependencies to their latest versions.
- Improved Building and Usage Documentation.

## 1.3.0

### ⚠️Breaking Changes

- `D2V()` object and `loads` now expect a path argument which may be `None` or a Path to a D2V object.
- `load` now uses the new `_get_d2v()` (explained below) on the input file path (unless it's already a D2V).

### Improvements

- New static method `_get_d2v()` has been created, which can be used to generate a D2V for an input video file by path.
- Path to the input D2V file is now stored inside the D2V object for referencing (`D2V.path`). As stated above,
  `D2V()` and `loads` will now ask for an optional Path.
- All public-facing class methods now enforce the type of the arguments, as well as checks if the file actually exists
  and is not a directory.
- Variables formatted into `D2V()`s repr is now also using their reprs as well. This improves its readability.
- Use consistent double-quotes throughout pyproject.toml.

## 1.2.0

### ⚠️Breaking Changes

- Add `load` and `loads` class methods to D2V class giving the ability to parse from a file or string.
- `D2V()` now expects a TextIO based object (e.g. `open(mode="rt")` or `StringIO`). You may wish to use the new
  `load` and `loads` commands instead if possible.

### Bug Fixes

- Cast flag file data as int, since it's an index to a file parsed into `D2V.videos`.

### Improvements

- Create this `HISTORY.md`, add in all previous missed releases with no history.
- Add the ability to select data from the D2V when reading from the cli `d2v` terminal command.
  E.g. `d2v vid.d2v settings.Aspect_Ratio`
- Add doc-string to cli `d2v` terminal command.
- Optimize the code that actually parses the D2V, making it more concise and readable.
- Update copyright years in the LICENSE file.

## 1.1.0

### New Features

- Once installed, `d2v` will be an available terminal command. This can be used to read or parse then read a D2V file.

### Improvements

- Replace setuptools/setup.py with Poetry for a much better dependency and package management environment.
- Delete now unnecessary bash scripts I used to use starting with v1.0.1 for setuptools usage (setuptools was annoying).
- D2V class moved from `__init__.py` to `d2v.py`. It's generally recommended for the init file to only have needed code
  on import only, it's also generally less readable at a glance where the class was. It's still imported and set in
  `__ALL__` so that `from pyd2v import D2V` is still possible.
- General improvements and to the README information and structure.

## 1.0.4

### Improvements

- Cast flag position, skip, vob, and cell data to integers.

## 1.0.3

### Bug fixes

- Add a check for the final flag, `ff` terminator. This verifies end-of-stream. It's not an actual flag, so it should
  not be parsed as one either (or even kept).

## 1.0.2

### Improvements

- Parse data info and flags at the bit-level.

## 1.0.1

### Bug fixes

- Remove int parsing on Location values; they are HEX, not int.

## 1.0.0

- Initial release.


pyd2v/__main__.py
from pathlib import Path
from typing import Optional

import jsonpickle

from pyd2v import D2V
import click


@click.command(context_settings=dict(
    help_option_names=["-?", "-h", "--help"],
    max_content_width=116,  # max PEP8 line-width, -4 to adjust for initial indent
))
@click.argument("file", type=Path)
@click.argument("key", type=str, required=False)
@click.option("-j", "--json", is_flag=True, default=False,
              help="Return D2V data as a JSON object.")
@click.option("-p", "--pretty", is_flag=True, default=False,
              help="Return D2V data as a prettified object.")
def main(file: Path, key: Optional[str], json: bool, pretty: bool):
    """
    \b
    Access a D2V file via CLI.

    \b
    Version 1.3.1
    Copyright (c) 2020-2023 rlaphoenix — MIT License
    For more information see the project at: https://github.com/rlaphoenix/pyd2v

    \b
    [FILE] is the path to a D2V object you wish to parse.
    [KEY] is an optional object-path you may wish to use to access data from the D2V.
          e.g. `settings.Aspect_Ratio` will get you the Aspect Ratio.
    """
    if not isinstance(file, Path) or not file:
        raise click.BadParameter("File is an invalid value. Must be a pathlib.Path object.", param_hint="file")
    if not file.exists():
        raise click.BadParameter("File path does not exist.", param_hint="file")
    if not file.is_file():
        raise click.BadParameter("File path is to a directory, not a file.", param_hint="file")
    d2v = D2V.load(file)
    if key:
        for path in key.split("."):
            if isinstance(d2v, dict):
                d2v = d2v[path]
            else:
                d2v = getattr(d2v, path)
    if json:
        print(jsonpickle.encode(
            d2v,
            unpicklable=True,
            indent=2 if pretty else None,
            separators=(",", ": ") if pretty else None
        ))
        return
    print(d2v)


pyd2v/d2v.py
import shutil
import subprocess
from io import StringIO
from pathlib import Path
from typing import TextIO


class D2V:
    """DGIndex D2V project file."""

    def __init__(self, f: TextIO, path: Path = None):
        """
        Parse a d2v.
        :param f: Text IO object of a d2v to be parsed.
        :param path: Path object of the D2V file being parsed.
        :raises ValueError: if parsing fails
        """
        if not isinstance(path, Path) and path is not None:
            raise TypeError(f"D2V path must be a Path object (or None), not {type(path)}")
        if path is not None:
            if not path.exists():
                raise ValueError("D2V path does not exist")
            if not path.is_file():
                raise ValueError("D2V path is to a directory, not a file")
        self.path = path
        self.version = None
        self.videos = None
        self.settings = None
        self.data = None
        self.data_type = None

        """ Header Data """

        self.version = f.readline().strip()
        if not self.version.startswith("DGIndexProjectFile"):
            raise ValueError(f"Expected Version Header, received '{self.version}'")
        self.version = int(self.version[18:])  # strip "DGIndexProjectFile"
        self.videos = [f.readline().strip() for _ in range(int(f.readline().strip()))]

        if len(f.readline().strip()) > 0:
            raise ValueError("Unexpected data after reading Header's Video List.")

        """ Settings Data """

        self.settings = {}
        while True:
            line = f.readline().strip()
            if len(line) == 0:
                break
            key, value = line.split("=", maxsplit=2)
            if key in ["Stream_Type", "MPEG_Type", "iDCT_Algorithm", "YUVRGB_Scale", "Field_Operation"]:
                value = int(value)
            if key in ["MPEG2_Transport_PID", "Transport_Packet_Size", "Luminance_Filter", "Clipping"]:
                value = list(map(int, value.split(",")))
            if key == "MPEG2_Transport_PID":
                value = {"Video": value[0], "Audio": value[1], "PCR": value[2]}
            elif key == "Luminance_Filter":
                value = {"Gamma": value[0], "Offset": value[1]}
            elif key == "Clipping":
                value = list(map(int, value))
            elif key == "Aspect_Ratio":
                if "," in value:
                    value = [(x if ":" in x else float(x)) for x in value.split(",")]
                elif ":" not in value:
                    value = float(value)
            elif key == "Picture_Size":
                value = list(map(int, value.lower().split("x")))
            elif key == "Frame_Rate":
                rate, num_den = value.split(" ")
                num_den = list(map(int, num_den.strip("()").split("/")))
                value = [rate, num_den]
            elif key == "Location":
                start_file, start_offset, end_file, end_offset = value.split(",")
                value = {
                    "StartFile": start_file,
                    "StartOffset": start_offset,
                    "EndFile": end_file,
                    "EndOffset": end_offset
                }
            self.settings[key] = value

        """ Flag Data """

        self.data = []
        while True:
            line = f.readline().strip()
            if len(line) == 0:
                break
            line = line.split(" ", maxsplit=7)
            data = {
                "info": bin(int(line[0], 16))[2:].zfill(8),
                "matrix": line[1],
                "file": int(line[2]),
                "position": int(line[3]),
                "skip": int(line[4]),
                "vob": int(line[5]),
                "cell": int(line[6]),
                "flags": line[7].split(" ")
            }
            data["info"] = {
                "data_line_signal": data["info"][0],
                "part_of_closed_gop": data["info"][1] == "1",
                "part_of_progressive_sequence": data["info"][2] == "1",
                "first_picture_of_new_gop": data["info"][3] == "1",
                "reserved": data["info"][4:]
            }
            if data["flags"][-1].lower() == "ff":
                # ff == end of stream indicator, not an actual frame data byte
                data["flags"] = data["flags"][0:-1]
            data["flags"] = [{
                "require_previous_gop": bit[0] == "0",
                "progressive_frame": bit[1] == "1",
                "picture_coding_type": {
                    "00": "reserved",
                    "01": "I",
                    "10": "P",
                    "11": "B"
                }[bit[2] + bit[3]],
                "reserved_bits": bit[4] + bit[5],
                "tff": bit[6] == "1",
                "rff": bit[7] == "1"
            } for bit in [bin(int(flag, 16))[2:].zfill(8) for flag in data["flags"]]]
            self.data.append(data)
        self.data_type = f.readline()[10:]

    def __repr__(self):
        return f"<D2V path={self.path!r} version={self.version!r}, data_type={self.data_type!r}, " \
               f"settings={self.settings!r}>"

    @classmethod
    def loads(cls, data: str, path: Path = None):
        """
        Parse a D2V from a blob of string data.

        The optional path argument is available to store the file location of
        the D2V data inside the new D2V object under `D2V.path` class variable.
        It must be a Path object, or otherwise None.
        """
        if not isinstance(path, Path) and path is not None:
            raise TypeError(f"path must be a Path object (or None), not {type(path)}")
        return cls(StringIO(data), path)

    @classmethod
    def load(cls, file: Path):
        """
        Parse a D2V from a file.

        If file is not a D2V path, it will assume it's a path to a video file and
        will generate an optimal D2V from the video file. It will also demux said
        file for D2V generation if it's in a container (e.g., mp4, mkv).

        The generated D2V will be next to the input file, and the direct path will
        be stored within the new D2V object in the `D2V.path` class variable.
        """
        if not isinstance(file, Path):
            raise TypeError(f"file must be a Path object, not {type(file)}")
        if not file.exists():
            raise ValueError("file path does not exist")
        if not file.is_file():
            raise ValueError("file path is to a directory, not a file")
        file = cls._get_d2v(file)
        return cls(file.open(mode="r", encoding="utf8"), file)

    @staticmethod
    def _get_d2v(file_path: Path, idct_algo=5, field_op=2, yuv_to_rgb=1) -> Path:
        """
        Index an optimal D2V file using DGIndex.
        Unix systems are supported as long as Wine is installed.

        Note, It will demux the video track from any containers as DGIndex does
        not support most container formats, and it's generally advised to do so
        even if it does support a specific container.

        The default extra arguments that gets passed to DGIndex is generally
        recommended to be untouched. Especially field_op! Please do not change
        these unless you know what they do. However, you may need to change
        yuv_to_rgb depending on you're source, but most MPEGs should be fine
        set to 1 (PC Scale).
        """
        is_vob = file_path.suffix.lower() == ".vob"
        d2v_path = file_path.with_suffix(".d2v")
        if d2v_path.is_file():
            return d2v_path

        # demux the mpeg stream if not a .VOB or .MPEG file
        demuxed_ext = [".mpeg", ".mpg", ".m2v", ".vob"]
        vid_path = next((x for x in map(file_path.with_suffix, demuxed_ext) if x.exists()), None)
        if not vid_path:
            vid_path = file_path.with_suffix(demuxed_ext[0])
            mkvextract = shutil.which("mkvextract")
            if not mkvextract:
                raise RuntimeError(
                    "Executable 'mkvextract' not found, but is needed for the provided file.\n"
                    "Install MKVToolNix and make sure it's binaries are in the environment path."
                )
            subprocess.run([
                mkvextract,
                file_path.name,
                # TODO: This assumes the track with track-id of 0 is the video. Use pymediainfo to get
                #       first video track's Track ID. For now assuming 0 will have to do.
                "tracks", f"0:{vid_path.name}"
            ], cwd=file_path.parent, check=True)

        # use dgindex to generate a d2v file for the demuxed track
        dgindex = shutil.which("dgindex")
        if not dgindex:
            raise RuntimeError(
                "Executable 'dgindex' not found, but is needed for the provided file.\n"
                "Add dgindex.exe to your environment path. Ensure the executable is named `dgindex.exe`."
            )
        is_unix = dgindex.startswith("/")
        if is_unix:
            # required to do it this way for whatever reason. Directly calling it sometimes fails.
            dgindex = ["wine", "start", "/wait", "Z:" + dgindex]
        else:
            dgindex = [dgindex]
        subprocess.run(
            dgindex + [
                "-ai" if is_vob else "-i", vid_path.name,
                "-ia", str(idct_algo),  # iDCT Algorithm, 5=IEEE-1180 Reference
                "-fo", str(field_op),  # Field Operation, 2=Ignore Pulldown Flags
                "-yr", str(yuv_to_rgb),  # YUV->RGB, 1=PC Scale
                "-om", "0",  # Output Method, 0=None (just d2v)
                "-hide", "-exit",  # start hidden and exit when saved
                "-o", file_path.stem
            ],
            cwd=file_path.parent,
            check=True
        )

        # replace the Z:\bla\bla paths to /bla/bla unix paths, if on a unix system.
        # This is needed simply due to how d2vsource loads the video files. On linux it doesn't use wine,
        # so Z:\ paths obviously won't exist.
        if is_unix:
            with open(d2v_path, "rt", encoding="utf8") as f:
                d2v_content = f.read().splitlines()
            d2v_content = [(x[2:].replace("\\", "/") if x.startswith("Z:\\") else x) for x in d2v_content]
            with open(d2v_path, "wt", encoding="utf8") as f:
                f.write("\n".join(d2v_content))

        return d2v_path
